common/types/encoding/da.go

```diff
@@ -0,0 +1,232 @@
+	var numL1Messages uint64
+	for _, block := range c.Blocks {
+		numL1MessagesInBlock := block.NumL1Messages(totalL1MessagePoppedBefore)
+		numL1Messages += numL1MessagesInBlock
+		totalL1MessagePoppedBefore += numL1MessagesInBlock
+	}
+	// TODO: cache results
+	return numL1Messages
+}
+
+// ConvertTxDataToRLPEncoding transforms []*TransactionData into []*types.Transaction.
+func ConvertTxDataToRLPEncoding(txData *types.TransactionData) ([]byte, error) {
+	data, err := hexutil.Decode(txData.Data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode txData.Data: %s, err: %w", txData.Data, err)
+	}
+
+	var tx *types.Transaction
+	switch txData.Type {
+	case types.LegacyTxType:
+		tx = types.NewTx(&types.LegacyTx{
+			Nonce:    txData.Nonce,
+			To:       txData.To,
+			Value:    txData.Value.ToInt(),
+			Gas:      txData.Gas,
+			GasPrice: txData.GasPrice.ToInt(),
+			Data:     data,
+			V:        txData.V.ToInt(),
+			R:        txData.R.ToInt(),
+			S:        txData.S.ToInt(),
+		})
+
+	case types.AccessListTxType:
+		tx = types.NewTx(&types.AccessListTx{
+			ChainID:    txData.ChainId.ToInt(),
+			Nonce:      txData.Nonce,
+			To:         txData.To,
+			Value:      txData.Value.ToInt(),
+			Gas:        txData.Gas,
+			GasPrice:   txData.GasPrice.ToInt(),
+			Data:       data,
+			AccessList: txData.AccessList,
+			V:          txData.V.ToInt(),
+			R:          txData.R.ToInt(),
+			S:          txData.S.ToInt(),
+		})
+
+	case types.DynamicFeeTxType:
+		tx = types.NewTx(&types.DynamicFeeTx{
+			ChainID:    txData.ChainId.ToInt(),
+			Nonce:      txData.Nonce,
+			To:         txData.To,
+			Value:      txData.Value.ToInt(),
+			Gas:        txData.Gas,
+			GasTipCap:  txData.GasTipCap.ToInt(),
+			GasFeeCap:  txData.GasFeeCap.ToInt(),
+			Data:       data,
+			AccessList: txData.AccessList,
+			V:          txData.V.ToInt(),
+			R:          txData.R.ToInt(),
+			S:          txData.S.ToInt(),
+		})
+
+	case types.L1MessageTxType:
+	default:
+		return nil, fmt.Errorf("unsupported tx type: %v", txData.Type)
+	}
+
+	rlpTxData, err := tx.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal binary of the tx: %+v, err: %w", tx, err)
+	}
+
+	return rlpTxData, nil
+}
+
+// GetCrcMax calculates the maximum row consumption of crc.
+func (c *Chunk) GetCrcMax() uint64 {
+	// Map sub-circuit name to row count
+	crc := make(map[string]uint64)
+
+	// Function to accumulate row consumption
+	add := func(rowConsumption *types.RowConsumption) {
+		if rowConsumption == nil {
+			return
+		}
+		for _, subCircuit := range *rowConsumption {
+			crc[subCircuit.Name] += subCircuit.RowNumber
+		}
+	}
+
+	// Function to find maximum row consumption
+	max := func() uint64 {
+		var maxVal uint64
+		for _, value := range crc {
+			if value > maxVal {
+				maxVal = value
+			}
+		}
+		return maxVal
+	}
+
+	// Iterate over blocks, accumulate row consumption
+	for _, block := range c.Blocks {
+		if block.RowConsumption != nil {
```

Comment:
It's better to throw an error when it's `nil` (like the original implementation). That way it's easy to detect if `rollup-relayer` is connected to a node with ccc disabled.

Reply:
added in [3337b05](https://github.com/scroll-tech/scroll/pull/1178/commits/3337b055f920fb9e445f9f713e42e42ea648850b)